File: WindowsMobileClient/Windows Mobile Client/Menus/EditContextMenu.vb
Imports System.ComponentModel
Imports activiser.Library.Forms

<ComponentModel.DesignTimeVisible(True)> _
Public Class EditContextMenu : Inherits ContextMenuBase

    Public Sub New()
        MyBase.New(ContextMenuType.Normal)
    End Sub

    Private _showCall As Boolean = True
    Public Property ShowCall() As Boolean
        Get
            Return _showCall
        End Get
        Set(ByVal value As Boolean)
            _showCall = value
        End Set
    End Property

    Protected Overrides Sub OnPopup(ByVal e As System.EventArgs)
        MyBase.OnPopup(e)
        Dim enableCuts As Boolean
        Dim enablePastes As Boolean
        Dim enablePhone As Boolean
        Dim enableSelectAll As Boolean
        'Dim enableMail As Boolean

        ' probly not the prettiest, but looks slightly better than a nested if...

        If TestTextBox(TryCast(Me.SourceControl, TextBox), enableCuts, enablePastes, enablePhone, enableSelectAll) Then
            'do nothing
        ElseIf TestNumberPicker(TryCast(Me.SourceControl, NumberPicker), enableCuts, enablePastes, enablePhone, enableSelectAll) Then
            'do nothing
        ElseIf TestComboBox(TryCast(Me.SourceControl, ComboBox), enableCuts, enablePastes, enablePhone, enableSelectAll) Then
            'do nothing
        ElseIf TestDateTimePicker(TryCast(Me.SourceControl, DateTimePicker), enableCuts, enablePastes, enablePhone, enableSelectAll) Then
            'do nothing
        Else
            TestLabel(TryCast(Me.SourceControl, Label), enableCuts, enablePastes, enablePhone, enableSelectAll)
        End If

#If WINDOWSMOBILE Then
        If _showCall AndAlso HavePhone() Then
            If Not Me.MenuItems.Contains(CallMI) Then
                Me.MenuItems.Add(CallMI)
            End If
            CallMI.Enabled = enablePhone
        Else
            If Me.MenuItems.Contains(CallMI) Then
                Me.MenuItems.Remove(CallMI)
            End If
        End If
#End If

        CutMI.Enabled = enableCuts
        ClearMI.Enabled = enableCuts
        PasteMI.Enabled = enablePastes
        SelectAllMI.Enabled = enableSelectAll
    End Sub

    Private Shared Function TestTextBox(ByVal tb As TextBox, ByRef enableCuts As Boolean, ByRef enablePastes As Boolean, ByRef enablePhone As Boolean, ByRef enableSelectAll As Boolean) As Boolean
        If tb Is Nothing Then Return False
        enableCuts = tb.Enabled AndAlso Not tb.ReadOnly
        enablePastes = enableCuts
        enableSelectAll = True
#If WINDOWSMOBILE Then
        enablePhone = CanPhone(tb)
#End If
        Return True
    End Function

    Private Shared Function TestNumberPicker(ByVal nb As NumberPicker, ByRef enableCuts As Boolean, ByRef enablePastes As Boolean, ByRef enablePhone As Boolean, ByRef enableSelectAll As Boolean) As Boolean
        If nb Is Nothing Then Return False
        enableCuts = nb.Enabled AndAlso Not nb.ReadOnly
        enablePastes = enableCuts
        enableSelectAll = False
#If WINDOWSMOBILE Then
        enablePhone = CanPhone(nb.Text)
#End If       
        Return True
    End Function

    Private Shared Function TestComboBox(ByVal cb As ComboBox, ByRef enableCuts As Boolean, ByRef enablePastes As Boolean, ByRef enablePhone As Boolean, ByRef enableSelectAll As Boolean) As Boolean
        If cb Is Nothing Then Return False
        enableCuts = cb.Enabled AndAlso cb.DropDownStyle = ComboBoxStyle.DropDown
        enablePastes = cb.Enabled
        enableSelectAll = False
#If WINDOWSMOBILE Then
        enablePhone = CanPhone(cb.Text)
#End If       
        Return True
    End Function

    Private Shared Function TestDateTimePicker(ByVal dt As DateTimePicker, ByRef enableCuts As Boolean, ByRef enablePastes As Boolean, ByRef enablePhone As Boolean, ByRef enableSelectAll As Boolean) As Boolean
        If dt Is Nothing Then Return False
        enableCuts = False
        enablePastes = dt.Enabled
        enablePhone = False
        enableSelectAll = False
        Return True
    End Function

    Private Shared Function TestLabel(ByVal lb As Label, ByRef enableCuts As Boolean, ByRef enablePastes As Boolean, ByRef enablePhone As Boolean, ByRef enableSelectAll As Boolean) As Boolean
        If lb Is Nothing Then Return False
        enableCuts = False
        enablePastes = False
        enableSelectAll = False
#If WINDOWSMOBILE Then
        enablePhone = CanPhone(lb.Text)
#End If
        Return True
    End Function
End Class

File: Library/activiser.Library.Forms/EditContextMenu.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.ComponentModel;
using System.Windows.Forms;

namespace activiser.Library.Forms
{
    [DesignTimeVisible(true), DesignerCategory("Component")]
    public class EditContextMenu : ContextMenu
    {
        EditContextMenu()
            : base()
        {
            MenuItems.Add(ContextMenuBase.CutMI);
            MenuItems.Add(ContextMenuBase.CopyMI);
            MenuItems.Add(ContextMenuBase.PasteMI);
            MenuItems.Add(ContextMenuBase.ClearMI);
            MenuItems.Add(ContextMenuBase.SelectAllMI);

            if (Phone.HavePhone())
            {
                MenuItems.Add(ContextMenuBase.CallMI);
            }
        }
        //private bool _showCall = Phone.HavePhone();
        //public bool ShowCall
        //{
        //    get { return _showCall; }
        //    set
        //    {
        //        _showCall = value;
        //        if (_showCall && Phone.HavePhone() && !this.MenuItems.Contains(ContextMenuBase.CallMI))
        //        {
        //            this.MenuItems.Add(ContextMenuBase.CallMI);
        //        }
        //        else
        //        {
        //            if (this.MenuItems.Contains(ContextMenuBase.CallMI)) this.MenuItems.Remove(ContextMenuBase.CallMI);
        //        }
        //    }
        //}

        protected override void OnPopup(System.EventArgs e)
        {
            base.OnPopup(e);
            //bool enableCuts = false;
            //bool enablePastes = false;
            //bool enablePhone = false;
            //bool enableSelectAll = false;
            ////Dim enableMail As Boolean 

            //if (!TestTextBox(this.SourceControl as TextBox, ref enableCuts, ref enablePastes, ref enablePhone, ref enableSelectAll))
            //    if (!TestNumberPicker(this.SourceControl as NumberPicker, ref enableCuts, ref enablePastes, ref enablePhone, ref enableSelectAll))
            //        if (!TestComboBox(this.SourceControl as ComboBox, ref enableCuts, ref enablePastes, ref enablePhone, ref enableSelectAll))
            //            if (!TestDateTimePicker(this.SourceControl as DateTimePicker, ref enableCuts, ref enablePastes, ref enablePhone, ref enableSelectAll))
            //                TestLabel(this.SourceControl as Label, ref enableCuts, ref enablePastes, ref enablePhone, ref enableSelectAll);

            //ContextMenuBase.CutMI.Enabled = enableCuts;
            //ContextMenuBase.ClearMI.Enabled = enableCuts;
            //ContextMenuBase.PasteMI.Enabled = enablePastes;
            //ContextMenuBase.SelectAllMI.Enabled = enableSelectAll;

            //if (ShowCall) ContextMenuBase.CallMI.Enabled = enablePhone;
        }

        private static bool TestTextBox(TextBox tb, ref bool enableCuts, ref bool enablePastes, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (tb == null)
                return false;
            enableCuts = tb.Enabled && !tb.ReadOnly;
            enablePastes = enableCuts;
            enableSelectAll = true;
            enablePhone = Phone.CanPhone(tb);
            return true;
        }

        private static bool TestNumberPicker(NumberPicker nb, ref bool enableCuts, ref bool enablePastes, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (nb == null)
                return false;
            enableCuts = nb.Enabled && !nb.ReadOnly;
            enablePastes = enableCuts;
            enableSelectAll = true;
            if (string.IsNullOrEmpty(nb.SelectedText))
                enablePhone = Phone.CanPhone(nb.SelectedText);
            else
                enablePhone = Phone.CanPhone(nb.ToString());

            return true;
        }

        private static bool TestComboBox(ComboBox cb, ref bool enableCuts, ref bool enablePastes, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (cb == null)
                return false;
            enableCuts = cb.Enabled && cb.DropDownStyle == ComboBoxStyle.DropDown;
            enablePastes = cb.Enabled;
            enableSelectAll = false;
            enablePhone = Phone.CanPhone(cb.Text);
            return true;
        }

        private static bool TestDateTimePicker(DateTimePicker dt, ref bool enableCuts, ref bool enablePastes, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (dt == null)
                return false;
            enableCuts = false;
            enablePastes = dt.Enabled;
            enablePhone = false;
            enableSelectAll = false;
            return true;
        }

        private static bool TestLabel(Label lb, ref bool enableCuts, ref bool enablePastes, ref bool enablePhone, ref bool enableSelectAll)
        {
            if (lb == null)
                return false;
            enableCuts = false;
            enablePastes = false;
            enableSelectAll = false;
            enablePhone = Phone.CanPhone(lb.Text);
            return true;
        }
    } 
}
